2024/04/day04.py

```diff
@@ -0,0 +1,70 @@
+def read_input(file_name: str) -> list[str]:
+    with open(file_name) as f:
+        return f.readlines()
+
+def is_word(map_: list[str], word: str, row: int, column: int, row_diff: int, column_diff: int) -> bool:
+    if row < 0 or len(map_) <= row:
+        return False
+    if column < 0 or len(map_[row]) <= column:
+        return False
+    if len(word) == 0:
+        return False
+    if map_[row][column] != word[0]:
+        return False
+    if len(word) == 1:
+        return True
+    next_row = row + row_diff
+    next_column = column + column_diff
+    return is_word(map_, word[1:], next_row, next_column, row_diff, column_diff)
+
+def count_xmas(map_: list[str]) -> int:
+    """
+    mapをn * mとすると、
+    時間計算量: O(nm)。 「すべての位置」に対して、O(1)のXMASを探索する操作を9回行うため。
+    空間計算量: O(1)。
+    """
+    xmas_count = 0
+    for row in range(len(map_)):
+        for column in range(len(map_[row])):
+            for row_diff in [-1, 0, 1]:
+                for column_diff in [-1, 0, 1]:
+                    if is_word(map_, 'XMAS', row, column, row_diff, column_diff):
+                        xmas_count += 1
```

Comment:
4重ループは見たことがないので、さすがにネストが深すぎるのではという気がします。9方向を表す変数を定義してそれでループを回せば3重に抑えられそうです。

```suggestion
    moves = [(-1, -1), (-1, 0), ...] // 9方向
    ...
            for row_diff, column_diff in moves:
                if is_word(map_, 'XMAS', row, column, row_diff, column_diff):
                    xmas_count += 1
```